src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/ReferingToObjectAndReassigningItInTheSameStatement.cs

```diff
@@ -0,0 +1,96 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeQuality.Analyzers.QualityGuidelines
+{
+    /// <summary>
+    /// ReferenceChanged: Prevent objects from being referenced in statements where they are reassigned
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+
+    public sealed class ReferingToObjectAndReassigningItInTheSameStatement : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "ReferenceChanged";
+
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(MicrosoftQualityGuidelinesAnalyzersResources.ReferingToObjectAndReassigningItInTheSameStatementTitle), MicrosoftQualityGuidelinesAnalyzersResources.ResourceManager, typeof(MicrosoftQualityGuidelinesAnalyzersResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(MicrosoftQualityGuidelinesAnalyzersResources.ReferingToObjectAndReassigningItInTheSameStatementMessage), MicrosoftQualityGuidelinesAnalyzersResources.ResourceManager, typeof(MicrosoftQualityGuidelinesAnalyzersResources));
+
+        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Usage,
+            DiagnosticHelpers.DefaultDiagnosticSeverity,
+            isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            helpLinkUri: null);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterOperationAction(AnalyzeAssignment, OperationKind.SimpleAssignment);
+        }
+
+        private void AnalyzeAssignment(OperationAnalysisContext context)
+        {
+            var assignmentOperation = (ISimpleAssignmentOperation)context.Operation;
+
+            // Check if there are more then one assignment in a statement
+            if (!(assignmentOperation.Target is IMemberReferenceOperation operationTarget))
+            {
+                return;
+            }
+
+            // This analyzer makes sense only for reference type objects
+            if (operationTarget.Instance?.Type.IsValueType == true)
+            {
+                return;
+            }
+
+            // Search for object equal to operationTarget.Instance further in assignment chain
+            bool isViolationFound = false;
+            string violatingObjectName;
+            if (operationTarget.Instance is ILocalReferenceOperation localInstance)
+            {
+                violatingObjectName = localInstance.Local.Name;
+                isViolationFound = AnalyzeMemberAssignment(assignmentOperation, localInstance, (a, b) => a.Local == b.Local);
```

Comment:
Nit: rename to `AnalyzeAssignmentToMember` as this seems to indicate the passed in operation is an `IMemberAssignmentOperation`